### PR TITLE
Chapter 12

### PR DESCRIPTION
**Pull request by mirulzaei**

### CODE REVIEW

**Comment by mirulzaei** [comment]
ready

File: app/assets/stylesheets/account_activations.scss [hunk 1]
```diff
@@ -0,0 +1,3 @@
+// Place all the styles related to the AccountActivations controller here.
+// They will automatically be included in application.css.
+// You can use Sass (SCSS) here: https://sass-lang.com/
```
Comment (saobv-2223) [inline app/assets/stylesheets/account_activations.scss:3]: remove file that never used

File: app/assets/stylesheets/password_resets.scss [hunk 2]
```diff
@@ -0,0 +1,3 @@
+// Place all the styles related to the PasswordResets controller here.
+// They will automatically be included in application.css.
+// You can use Sass (SCSS) here: https://sass-lang.com/
```
Comment (saobv-2223) [inline app/assets/stylesheets/password_resets.scss]: remove file that never used

File: app/controllers/account_activations_controller.rb [hunk 3]
```diff
@@ -0,0 +1,2 @@
+class AccountActivationsController < ApplicationController
+end
```
Comment (saobv-2223) [inline app/controllers/account_activations_controller.rb]: remove file that never used

File: app/controllers/password_resets_controller.rb [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+  def user_params
+    params.require(:user).permit :password, :password_confirmation
+  end
+
+  def check_expiration
+    return unless @user.password_reset_expired?
+    flash[:danger] = "Password reset has expired."
+    redirect_to new_password_reset_url
+  end
+end
```
Comment (saobv-2223) [inline app/controllers/password_resets_controller.rb]: add endline

File: app/helpers/account_activations_helper.rb [hunk 5]
```diff
@@ -0,0 +1,2 @@
+module AccountActivationsHelper
+end
```
Comment (saobv-2223) [inline app/helpers/account_activations_helper.rb]: remove file that never used

File: app/helpers/password_resets_helper.rb [hunk 6]
```diff
@@ -0,0 +1,2 @@
+module PasswordResetsHelper
+end
```
Comment (saobv-2223) [inline app/helpers/password_resets_helper.rb]: remove file that never used

File: app/mailers/user_mailer.rb [hunk 7]
```diff
@@ -0,0 +1,12 @@
...
+  def account_activation(user)
+    @user = user
+    mail to: user.email, subject: "Account activation"
+  end
+
+  def password_reset
+    @user = user
+    mail to: user.email, subject: "Password reset"
+  end
+end
```
Comment (saobv-2223) [inline app/mailers/user_mailer.rb]: add endline

File: app/mailers/user_mailer.rb [hunk 8]
```diff
@@ -0,0 +1,12 @@
+class UserMailer < ApplicationMailer
+
+  def account_activation(user)
```
Comment (saobv-2223) [inline app/mailers/user_mailer.rb]: remove ()

File: app/models/user.rb [hunk 9]
```diff
@@ -43,9 +43,41 @@ def forgets
     update_attribute :remember_digest, nil
   end
 
+  def authenticated? attribute, token
+    digest = send "#{attribute}_digest"
+    return false unless digest
+    BCrypt::Password.new(digest).is_password? token
```
Comment (saobv-2223) [inline app/models/user.rb]:
```suggestion
    return false unless digest
    
    BCrypt::Password.new(digest).is_password? token
```

File: app/views/password_resets/edit.html.erb [hunk 10]
```diff
@@ -0,0 +1,17 @@
...
+      <%= hidden_field_tag :email, @user.email %>
+      <%= f.label :password %>
+      <%= f.password_field :password, class: "form-control" %>
+
+      <%= f.label :password_confirmation, "Confirmation" %>
+      <%= f.password_field :password_confirmation, class: "form-control" %>
+      <%= f.submit "Update password", class: "btn btn-primary" %>
+    <% end %>
+  </div>
+</div>
```
Comment (saobv-2223) [inline app/views/password_resets/edit.html.erb]: add endline

**Review by saobv-2223 — COMMENTED** [review]
Your PR is having 2 commits, merge to 1 commit, please

**Comment by mirulzaei** [comment]
ready

**Comment by manhnt-0960** [comment]
merged

**Pull request merged**